File: Ahorn/entities/strawberryJam/LoopBlock.jl
module CommunalHelperSJLoopBlock

using ..Ahorn, Maple

@mapdef Entity "CommunalHelper/SJ/LoopBlock" LoopBlock(x::Integer, y::Integer, width::Integer=16, height::Integer=16, edgeThickness::Integer=1, color::String="FFFFFF")

const placements = Ahorn.PlacementDict(
    "Loop Block (Strawberry Jam) (Communal Helper)" => Ahorn.EntityPlacement(
        LoopBlock,
        "rectangle"
    )
)

Ahorn.minimumSize(entity::LoopBlock) = 24, 24
Ahorn.resizable(entity::LoopBlock) = true, true

function Ahorn.selection(entity::LoopBlock)
    x, y = Ahorn.position(entity)
    width = Int(get(entity.data, "width", 16))
    height = Int(get(entity.data, "height", 16))

    return Ahorn.Rectangle(x, y, width, height)
end

const tileset = "objects/CommunalHelper/strawberryJam/loopBlock/tiles"

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::LoopBlock, room::Maple.Room)
    width = Int(get(entity.data, "width", 16))
    height = Int(get(entity.data, "height", 16))

    tileWidth = ceil(Int, width / 8)
    tileHeight = ceil(Int, height / 8)
    
    tileRands = Ahorn.getDrawableRoom(Ahorn.loadedState.map, room).fgTileStates.rands

    minSize = min(width, height) / 8
    edgeThickness = clamp(Int(get(entity.data, "edgeThickness", 1)), 1, floor((minSize - 1) / 2))

    # borrowed from Communal Helper
    color = tuple(Ahorn.argb32ToRGBATuple(parse(Int, String(get(entity.data, "color", "FFFFFF")), base=16))[1:3] ./ 255..., 1.0)

    tiles = Array{Bool}(undef, tileWidth, tileHeight)
    for i in 1:tileWidth, j in 1:tileHeight
        tiles[i, j] = (i <= edgeThickness || i > tileWidth - edgeThickness || j <= edgeThickness || j > tileHeight - edgeThickness)
    end

    # not a super elegant way of doing it
    for i in 1:tileWidth, j in 1:tileHeight
        if tiles[i, j] 
            up = get(tiles, (i, j - 1), false)
            down = get(tiles, (i, j + 1), false)
            left = get(tiles, (i - 1, j), false)
            right = get(tiles, (i + 1, j), false)
            upleft = get(tiles, (i - 1, j - 1), false)
            upright = get(tiles, (i + 1, j - 1), false)
            downleft = get(tiles, (i - 1, j + 1), false)
            downright = get(tiles, (i + 1, j + 1), false)

            tileChoice = mod(get(tileRands, (j, i), 0), 3)

            innerCorner = up && down && left && right
            full = innerCorner && upleft && upright && downleft && downright

            tx = 0
            ty = 0

            if full
                # used to be confusing, we are mapping [0, 1, 2, 3] to [0, 1, 0, 1] for tx, and [0, 0, 1, 1] for ty.
                # thanks vexatos
                r = mod(get(tileRands, (j, i), 0), 4)
                tx = mod(r, 2) + 1
                ty = div(r, 2) + 7
                tileChoice = 0
            elseif innerCorner
                tileChoice = 0
                if !downright
                    ty = 7
                elseif !downleft
                    tx = 3
                    ty = 7
                elseif !upright
                    ty = 8
                else 
                    tx = 3
                    ty = 8
                end
            else
                if !up && left && down && right
                    ty = 2
                elseif up && !left && down && right
                    tx = 3
                    ty = 2
                elseif up && left && !down && right
                    ty = 3
                elseif up && left && down && !right
                    tx = 3
                    ty = 3
                elseif !up && !left && down && right
                    ty = downright ? 0 : 4
                elseif !up && left && down && !right
                    tx = 3
                    ty = downleft ? 0 : 4
                elseif up && !left && !down && right
                    ty = upright ? 1 : 5
                elseif up && left && !down && !right
                    tx = 3
                    ty = upleft ? 1 : 5
                elseif up && down && !left && !right
                    tx = 3
                    ty = 6
                elseif !up && !down && left && right
                    ty = 6
                end
            end

            tx += tileChoice
            Ahorn.drawImage(ctx, tileset, (i - 1) * 8, (j - 1) * 8, tx * 8, ty * 8, 8, 8, tint=color)
        end
    end
end

end

File: Loenn/entities/strawberry_jam/loop_block.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")
local matrix = require("utils.matrix")

local loopBlock = {}

local defaultTexture = "objects/CommunalHelper/strawberryJam/loopBlock/tiles"

loopBlock.name = "CommunalHelper/SJ/LoopBlock"
loopBlock.minimumSize = {24, 24}

loopBlock.fieldInformation = {
    edgeThickness = {
        fieldType = "integer",
        minimumValue = 1
    },
    color = {
        fieldType = "color"
    },
    texture = {
        options = { defaultTexture },
        editable = true,
    }
}

loopBlock.placements = {
    {
        name = "loop_block",
        data = {
            width = 24,
            height = 24,
            edgeThickness = 1,
            color = "FFFFFF",
            texture = defaultTexture,
        }
    }
}

-- generates a 2d array of boolean that represent whether the block has a tile, at each index
local function generateTorusTiles(w, h, thickness)
    return matrix.fromFunction(
        function(i, _, _)
            local x, y = 1 + (i - 1) % w, 1 + math.floor((i - 1) / w)
            return x <= thickness or x > w - thickness
                or y <= thickness or y > h - thickness
        end,
        w, h
    )
end

function loopBlock.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 24, entity.height or 24
    local w, h = math.floor(width / 8), math.floor(height / 8)

    math.randomseed(x, y)

    local sprites = {}

    local tileset = entity.texture or defaultTexture
    local tiles = generateTorusTiles(w, h, entity.edgeThickness or 1)

    for i = 0, w - 1 do
        for j = 0, h - 1 do
            if tiles:get0(i, j, false) then
                local up, down = tiles:get0(i, j - 1, false), tiles:get0(i, j + 1, false)
                local left, right = tiles:get0(i - 1, j, false), tiles:get0(i + 1, j, false)
                local upleft, upright = tiles:get0(i - 1, j - 1, false), tiles:get0(i + 1, j - 1, false)
                local downleft, downright = tiles:get0(i - 1, j + 1, false), tiles:get0(i + 1, j + 1, false)

                local innerCorner = up and down and left and right
                local full = innerCorner and upleft and upright and downleft and downright

                local tx, ty = 0, 0

                if full then
                    tx, ty = 1 + math.random(0, 1), 7 + math.random(0, 1)
                elseif innerCorner then
                    if not downright then
                        tx, ty = 0, 7
                    elseif not downleft then
                        tx, ty = 3, 7
                    elseif not upright then
                        tx, ty = 0, 8
                    elseif not upleft then
                        tx, ty = 3, 8
                    end
                else
                    if (not up) and down and left and right then
                        tx, ty = 0, 2
                    elseif up and (not down) and left and right then
                        tx, ty = 0, 3
                    elseif up and down and (not left) and right then
                        tx, ty = 3, 2
                    elseif up and down and left and (not right) then
                        tx, ty = 3, 3
                    elseif (not up) and down and (not left) and right then
                        tx, ty = 0, downright and 0 or 4
                    elseif (not up) and down and left and (not right) then
                        tx, ty = 3, downleft and 0 or 4
                    elseif up and (not down) and (not left) and right then
                        tx, ty = 0, upright and 1 or 5
                    elseif up and (not down) and left and (not right) then
                        tx, ty = 3, upleft and 1 or 5
                    elseif (not up) and (not down) and left and right then
                        tx, ty = 0, 6
                    elseif up and down and (not left) and (not right) then
                        tx, ty = 3, 6
                    end

                    tx = tx + math.random(0, 2)
                end

                local tile = drawableSprite.fromTexture(tileset, entity)
                tile:useRelativeQuad(tx * 8, ty * 8, 8, 8)
                tile:addPosition(i * 8, j * 8)
                -- entity color is passed through the entity table,
                -- which we use as option table when calling drawableSprite.fromTexture

                table.insert(sprites, tile)
            end
        end
    end

    return sprites
end

function loopBlock.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 24, entity.height or 24
    return utils.rectangle(x, y, width, height)
end

return loopBlock
